Dokka plugin breaks task scanning

Extension Name: vscode-gradle
Extension Version: 3.12.3
OS Version:  macos 12.5.1
VSCode version: 1.70.2

When I add the plugin org.jetbrains.dokka, the gradle extension is not able to identify tasks properly. See the screenshots below. 

(I realise that the problem might be with the dokka plugin rather than with the gradle extension, but I have to start in this end, since this is where the problem shows up.)

1. Create a gradle project with gradle init. Choose groovy DSL, kotlin implemenation language, and library project.
2. Open the folder in VSCode and look at the gradle task list.
3. Add the dokka plugin to the gradle.build file
4. The gradle task list is not much shorter.

**Screenshots**
## Without dokka plugin
![without-dokka-plugin](https://user-images.githubusercontent.com/45693302/187149437-21f2d4d3-ee69-41bf-ad05-dd7726367077.png)

## With dokka plugin
![with-dokka-plugin](https://user-images.githubusercontent.com/45693302/187149526-575b4349-65b3-4719-9856-99040255a096.png)


**Output from "Gradle for Java"**
````
[info] Build file changed: /Users/xgumag/Documents/GU/projekt/aktiva_projekt/programmering/kutil/kutil/build.gradle
[info] Java Home: /usr/local/Cellar/openjdk@11/11.0.15/libexec/openjdk.jdk/Contents/Home
[info] JVM Args: --add-opens,java.base/java.util=ALL-UNNAMED,--add-opens,java.base/java.lang=ALL-UNNAMED,--add-opens,java.base/java.lang.invoke=ALL-UNNAMED,--add-opens,java.base/java.util=ALL-UNNAMED,--add-opens,java.prefs/java.util.prefs=ALL-UNNAMED,--add-opens,java.prefs/java.util.prefs=ALL-UNNAMED,--add-opens,java.base/java.nio.charset=ALL-UNNAMED,--add-opens,java.base/java.net=ALL-UNNAMED,--add-opens,java.base/java.util.concurrent.atomic=ALL-UNNAMED,-XX:MaxMetaspaceSize=256m,-XX:+HeapDumpOnOutOfMemoryError,-Xms256m,-Xmx512m,-Dfile.encoding=UTF-8,-Duser.country=SE,-Duser.language=sv,-Duser.variant
[info] Gradle User Home: /Users/xgumag/.gradle
[info] Gradle Version: 7.4.1
[error] FAILURE: Build failed with an exception.

* What went wrong:
Could not create task ':kutil:dokkaJavadoc'.
> Cannot change dependencies of dependency configuration ':kutil:dokkaJavadocPlugin' after it has been resolved.

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.

* Get more help at https://help.gradle.org
[info] Deprecated Gradle features were used in this build, making it incompatible with Gradle 8.0.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

See https://docs.gradle.org/7.4.1/userguide/command_line_interface.html#sec:command_line_warnings

CONFIGURE FAILED in 1s
[error] [gradle-server] The supplied build action failed with an exception.
[error] Error getting build for /Users/xgumag/Documents/GU/projekt/aktiva_projekt/programmering/kutil: The supplied build action failed with an exception.
[info] Found 0 tasks
````

--- a/gradle-plugin/src/main/java/com/microsoft/gradle/DefaultGradleTask.java
+++ b/gradle-plugin/src/main/java/com/microsoft/gradle/DefaultGradleTask.java
@@ -59,4 +59,8 @@ public class DefaultGradleTask implements Serializable, GradleTask {
 	public boolean getDebuggable() {
 		return debuggable;
 	}
+
+	public void setDebuggable(boolean debuggable) {
+		this.debuggable = debuggable;
+	}
 }

--- a/gradle-plugin/src/main/java/com/microsoft/gradle/GradlePlugin.java
+++ b/gradle-plugin/src/main/java/com/microsoft/gradle/GradlePlugin.java
@@ -19,6 +19,6 @@ public class GradlePlugin implements Plugin<Project> {
 
 	@Override
 	public void apply(Project project) {
-		registry.register(new GradleProjectModelBuilder());
+		registry.register(new GradleProjectModelBuilder(registry));
 	}
 }

--- a/gradle-plugin/src/main/java/com/microsoft/gradle/GradleProjectModelBuilder.java
+++ b/gradle-plugin/src/main/java/com/microsoft/gradle/GradleProjectModelBuilder.java
@@ -23,6 +23,7 @@ import java.util.Map;
 import java.util.Set;
 import java.util.TreeSet;
 import org.gradle.api.Project;
+import org.gradle.api.Task;
 import org.gradle.api.artifacts.Configuration;
 import org.gradle.api.artifacts.ConfigurationContainer;
 import org.gradle.api.artifacts.ResolvableDependencies;
@@ -41,7 +42,9 @@ import org.gradle.api.tasks.JavaExec;
 import org.gradle.api.tasks.TaskContainer;
 import org.gradle.api.tasks.testing.Test;
 import org.gradle.internal.classpath.ClassPath;
+import org.gradle.plugins.ide.internal.tooling.model.DefaultGradleProject;
 import org.gradle.tooling.provider.model.ToolingModelBuilder;
+import org.gradle.tooling.provider.model.ToolingModelBuilderRegistry;
 import org.gradle.util.GradleVersion;
 
 public class GradleProjectModelBuilder implements ToolingModelBuilder {
@@ -49,6 +52,11 @@ public class GradleProjectModelBuilder implements ToolingModelBuilder {
 	private static String MINIMAL_SUPPORTED_PLUGIN_CLOSURE_VERSION = "5.0";
 
 	private Set<GradleTask> cachedTasks = new HashSet<>();
+	private ToolingModelBuilderRegistry registry;
+
+	public GradleProjectModelBuilder(ToolingModelBuilderRegistry registry) {
+		this.registry = registry;
+	}
 
 	public boolean canBuild(String modelName) {
 		return modelName.equals(GradleProjectModel.class.getName());
@@ -56,7 +64,12 @@ public class GradleProjectModelBuilder implements ToolingModelBuilder {
 
 	public Object buildAll(String modelName, Project project) {
 		cachedTasks.clear();
-		GradleProjectModel rootModel = buildModel(project, project);
+		DefaultGradleProject gradleProject = (DefaultGradleProject) this.registry
+				.getBuilder("org.gradle.tooling.model.GradleProject").buildAll(modelName, project);
+		if (gradleProject == null) {
+			return null;
+		}
+		GradleProjectModel rootModel = buildModel(project, project.getName(), gradleProject);
 		// add task selectors for root project
 		Set<String> taskNames = new HashSet<>();
 		for (GradleTask existingTask : rootModel.getTasks()) {
@@ -80,8 +93,8 @@ public class GradleProjectModelBuilder implements ToolingModelBuilder {
 		return rootModel;
 	}
 
-	private GradleProjectModel buildModel(Project rootProject, Project project) {
-		if (rootProject == null || project == null) {
+	private GradleProjectModel buildModel(Project project, String rootProjectName, DefaultGradleProject gradleProject) {
+		if (project == null) {
 			return null;
 		}
 		ScriptHandler buildScript = project.getBuildscript();
@@ -94,14 +107,19 @@ public class GradleProjectModelBuilder implements ToolingModelBuilder {
 		List<String> plugins = getPlugins(project);
 		List<GradleClosure> closures = getPluginClosures(project);
 		List<GradleProjectModel> subModels = new ArrayList<>();
-		Map<String, Project> childProjects = project.getChildProjects();
-		for (Project childProject : childProjects.values()) {
-			GradleProjectModel subModel = buildModel(rootProject, childProject);
-			if (subModel != null) {
-				subModels.add(subModel);
+		for (DefaultGradleProject subDefaultGradleProject : gradleProject.getChildren()) {
+			// Query sub projects when both gradleProject and project contain them
+			Map<String, Project> childProjects = project.getChildProjects();
+			String projectName = subDefaultGradleProject.getName();
+			if (childProjects.keySet().contains(projectName)) {
+				GradleProjectModel subModel = buildModel(childProjects.get(projectName), rootProjectName,
+						subDefaultGradleProject);
+				if (subModel != null) {
+					subModels.add(subModel);
+				}
 			}
 		}
-		List<GradleTask> tasks = getGradleTasks(rootProject, project);
+		List<GradleTask> tasks = getGradleTasks(project, rootProjectName, gradleProject);
 		return new DefaultGradleProjectModel(project.getParent() == null, project.getProjectDir().getAbsolutePath(),
 				subModels, tasks, node, plugins, closures, scriptClasspaths);
 	}
@@ -202,30 +220,53 @@ public class GradleProjectModelBuilder implements ToolingModelBuilder {
 		return closures;
 	}
 
-	private List<GradleTask> getGradleTasks(Project rootProject, Project project) {
+	/**
+	 * get Task information from the given models. DefaultGradleProject is used to
+	 * get task list and Project is used to get debug information.
+	 *
+	 * @param project
+	 *            the given org.gradle.api.Project model
+	 * @param rootProjectName
+	 *            the root project name
+	 * @param gradleProject
+	 *            the given
+	 *            org.gradle.plugins.ide.internal.tooling.model.DefaultGradleProject
+	 *            model
+	 * @return the task list of the corresponding project
+	 */
+	private List<GradleTask> getGradleTasks(Project project, String rootProjectName,
+			DefaultGradleProject gradleProject) {
 		List<GradleTask> tasks = new ArrayList<>();
-		TaskContainer taskContainer = project.getTasks();
-		if (taskContainer instanceof TaskContainerInternal) {
-			TaskContainerInternal taskContainerInternal = (TaskContainerInternal) taskContainer;
-			taskContainerInternal.discoverTasks();
-			taskContainerInternal.realize();
-			taskContainerInternal.forEach(task -> {
-				String name = task.getName();
-				String group = task.getGroup() == null ? null : task.getGroup();
-				String path = task.getPath();
-				String projectName = task.getProject().getName();
-				String buildFile = task.getProject().getBuildscript().getSourceFile().getAbsolutePath();
-				String rootProjectName = rootProject.getName();
-				String description = task.getDescription() == null ? null : task.getDescription();
-				boolean debuggable = (task instanceof JavaExec) || (task instanceof Test);
-				GradleTask newTask = new DefaultGradleTask(name, group, path, projectName, buildFile, rootProjectName,
-						description, debuggable);
-				tasks.add(newTask);
-				cachedTasks.add(newTask);
-			});
-			return tasks;
+		gradleProject.getTasks().forEach((task) -> {
+			String group = task.getGroup() == null ? null : task.getGroup();
+			String description = task.getDescription() == null ? null : task.getDescription();
+			GradleTask newTask = new DefaultGradleTask(task.getName(), group, task.getPath(), gradleProject.getName(),
+					gradleProject.getBuildScript().getSourceFile().getAbsolutePath(), rootProjectName, description,
+					false);
+			tasks.add(newTask);
+			cachedTasks.add(newTask);
+		});
+		for (GradleTask gradleTask : tasks) {
+			// try to fetch debug information
+			TaskContainer taskContainer = project.getTasks();
+			if (taskContainer instanceof TaskContainerInternal) {
+				TaskContainerInternal taskContainerInternal = (TaskContainerInternal) taskContainer;
+				taskContainerInternal.discoverTasks();
+				taskContainerInternal.realize();
+				try {
+					Task task = taskContainerInternal.getByName(gradleTask.getName());
+					if ((task instanceof JavaExec || task instanceof Test)
+							&& (gradleTask instanceof DefaultGradleTask)) {
+						((DefaultGradleTask) gradleTask).setDebuggable(true);
+					}
+				} catch (Exception e) {
+					// for lazy tasks, `getByName()` will return an exception in some cases, we
+					// ignore them here
+					continue;
+				}
+			}
 		}
-		return Collections.emptyList();
+		return tasks;
 	}
 
 	private boolean isDeprecated(AccessibleObject object) {

--- a/gradle-server/src/main/java/com/github/badsyntax/gradle/handlers/GetBuildHandler.java
+++ b/gradle-server/src/main/java/com/github/badsyntax/gradle/handlers/GetBuildHandler.java
@@ -122,6 +122,9 @@ public class GetBuildHandler {
 					.setStandardOutput(standardOutputListener).setStandardError(standardErrorListener)
 					.setColorOutput(req.getShowOutputColors());
 			GradleProjectModel gradleModel = action.run();
+			if (gradleModel == null) {
+				throw new Exception("Error occurs in querying custom model.");
+			}
 			GradleProject project = getProjectData(gradleModel);
 			replyWithProject(project);
 		} catch (BuildCancelledException e) {

